src/Text/SlamSearch/Parser/Term.purs

```diff
@@ -0,0 +1,202 @@
+module Text.SlamSearch.Parser.Terms where
+
+import Text.SlamSearch.Parser.Values
+import Text.SlamSearch.Parser.Tokens
+import Text.SlamSearch.Parser.Utils
+
+import Text.Parsing.Parser
+import Text.Parsing.Parser.Combinators
+import Control.Apply
+import Control.Alt
+import Control.Alternative
+import Data.Tuple
+import Data.Either
+
+data SimplePredicate =
+  ContainsP Value
+  | EqP Value
+  | GtP Value
+  | GteP Value
+  | LteP Value
+  | LtP Value
+  | NeP Value
+  | LikeP Value
+
+instance showSimplePredicate :: Show SimplePredicate where
+  show a = case a of
+    ContainsP p -> "ContainsP(" <> show p <> ")"
+    EqP p -> "EqP(" <> show p <> ")"
+    GtP p -> "GtP(" <> show p <> ")"
+    GteP p -> "GteP(" <> show p <> ")"
+    LteP p -> "GteP(" <> show p <> ")"
+    LtP p -> "LtP(" <> show p <> ")"
+    NeP p -> "NeP(" <> show p <> ")"
+    LikeP p -> "LikeP(" <> show p <> ")"
+
+
+instance eqSimplePredicate :: Eq SimplePredicate where
+  (==) (ContainsP p) (ContainsP p') = p == p'
+  (==) (EqP p) (EqP p') = p == p'
+  (==) (GtP p) (GtP p') = p == p'
+  (==) (GteP p) (GteP p') = p == p'
+  (==) (LteP p) (LteP p') = p == p'
+  (==) (LtP p) (LtP p') = p == p'
+  (==) (NeP p) (NeP p') = p == p'
+  (==) (LikeP p) (LikeP p') = p == p'
+  (==) _ _ = false
+  (/=) a b = not $ a == b
+
+data Label =
+  Common String
+  | Meta String
+
+instance showLabel :: Show Label where
+  show a = case a of
+    Common p -> "Common(" <> show p <> ")"
+    Meta p -> "Meta(" <> show p <> ")"
+
+instance eqLabel :: Eq Label where
+  (==) (Common p) (Common p') = p == p'
+  (==) (Meta p) (Meta p') = p == p'
+  (==) _ _ = false
+  (/=) a b = not $ a == b
+
+data SearchTermSimple =
+  SimplePredicate SimplePredicate
+  | InfieldPredicate [Label] SimplePredicate
+
+instance showSearchTermSimpleEq :: Show SearchTermSimple where
+  show a = case a of
+    SimplePredicate p -> "SimplePredicate(" <> show p <> ")"
+    InfieldPredicate ls p -> "InfieldPredicate(" <> show ls <> "," <> show p <> ")"
+
+instance eqSearchTermSimple :: Eq SearchTermSimple where
+  (==) (SimplePredicate p) (SimplePredicate p') = p == p'
+  (==) (InfieldPredicate ls p) (InfieldPredicate ls' p') =
+    p == p' && ls == ls'
+  (==) _ _ = false
+  (/=) a b = not $ a == b
+
+data SearchTerm =
+  IncludeTerm SearchTermSimple
+  | ExcludeTerm SearchTermSimple
+
+instance searchTermEq :: Eq SearchTerm where
+  (==) (IncludeTerm t) (IncludeTerm t') = t == t'
+  (==) (ExcludeTerm t) (ExcludeTerm t') = t == t'
+  (==) _ _  = false
+  (/=) a b = not $ a == b
+
+instance showSearchTerm :: Show SearchTerm where
+  show a = case a of
+    IncludeTerm t -> "Include(" <> show t <> ")"
+    ExcludeTerm t -> "Exclude(" <> show t <> ")"
+
+data PredicateAndLabel =
```

Comment:
There are four data structures representing steps of parsing:

* _Token_ - string parsed, basic tokens
* _Value_ - values constructed from tokens, operators pushed through for next stesp
* _PredicateAndLabel_ - labels moved to ```L```, plus and minus to ```I``` for include and ```E``` for exclude. Operators and values composed to predicates.  
* _SearchTerm_ - labels bound to predicates. 

There is parser for each transition between states.